lychee-lib/src/client.rs

```diff
@@ -654,14 +661,32 @@ impl Client {
         let Ok(path) = uri.url.to_file_path() else {
             return ErrorKind::InvalidFilePath(uri.clone()).into();
         };
-        if !path.exists() {
-            return ErrorKind::InvalidFilePath(uri.clone()).into();
-        }
-        if self.include_fragments {
-            self.check_fragment(&path, uri).await
+        if path.exists() {
+            if self.include_fragments {
+                return self.check_fragment(&path, uri).await;
+            }
         } else {
-            Status::Ok(StatusCode::OK)
+            if self.check_extensions.is_empty() || path.extension().is_some() {
+                return ErrorKind::InvalidFilePath(uri.clone()).into();
+            }
+
+            // if the path has no file extension, try to append some
+            let mut path_buf = path.clone();
+            let mut found = false;
+            for ext in &self.check_extensions {
```

Comment:
Using the `found` flag and manually looping and breaking is not too nice. Instead try to use `.iter` and `find_map`.

Some ideas:

```rust
let check_extension_exists = |ext: &str| {
    let mut path_buf = path.clone();
    path_buf.set_extension(ext);
    if path_buf.exists() {
        Some(path_buf)
    } else {
        None
    }
};

let found_path = self.check_extensions.iter()
    .find_map(|ext| check_extension_exists(ext));

if let Some(path) = found_path {
  // ...
} else {}
```

Reply:
@thomas-zahner correct me if I am wrong, but don't we want to _stop_ at the first found extension? The suggested `map()` would not stop there, and in the best case do unnecessary work, in the worst case do unwanted work.

Reply:
[Here](https://github.com/lycheeverse/lychee/pull/1422/commits/976ba4b086b564badab92ef684cae7b44803446f#diff-c67c22c2f52695b3cdf26f7e6ba2fab2f49e4605941a7d87a005653d1c568e7aL654-L667) is my proposed code to address your concerns.

Reply:
Actually iterators are lazy and run only when it's really necessary. This is described in [the rust book](https://doc.rust-lang.org/book/ch13-02-iterators.html).

For example we can convert an infinite range of numbers into an iterator, map the numbers and take just a few of the numbers. Then the map only executes as many times as we take a result. In my code snippet above the `next()` consumes just one element so that the map only runs once.

```rust
    let numbers = (0..).into_iter().map(|x| x * 2).take(5);
    println!("{:?}", numbers.collect::<Vec<_>>());
```

Reply:
> Actually iterators are lazy and run only when it's really necessary. This is described in [the rust book](https://doc.rust-lang.org/book/ch13-02-iterators.html).
> 
> For example we can convert an infinite range of numbers into an iterator, map the numbers and take just a few of the numbers. Then the map only executes as many times as we take a result. In my code snippet above the `next()` consumes just one element so that the map only runs once.
> 
> ```rust
>     let iter = (0..).into_iter().map(|x| x * 2).take(5);
>     println!("{:?}", iter.collect::<Vec<_>>());
> ```

Okay, I didn't know that :-)

So now I'm torn: the code you showed that uses a `map()` looks less obvious than I'd like it to be, certainly more complicated than equivalent constructs in languages that I'm more familiar with than Rust. It may very well be more idiomatic Rust code than the `for` loop code that is in the current version of this PR. However, the latter version is much easier to parse and grok for me. So I am torn.

Reply:
Yeah, maybe my proposed iter solution is not perfect but normally iter is to be preferred over for loops. Maybe the iter solution could be simplified further. @mre what do you think?

Reply:
In general, I also prefer iterators, but in this case in particular, I'd lean towards the for-loop because of the early return. Now that the `found` flag got removed, it's gotten much more readable already. So I'm fine with the current version.

Reply:
So how does this PR proceed?

Reply:
Added a comment with a proposal.